notebooks/219-pytorch-object-detection-accuracy-check-and-quantization/219-pytorch-object-detection-accuracy-check-and-quantization.ipynb

```diff
@@ -0,0 +1,578 @@
+    "            np.bincount(stats[3].astype(np.int64), minlength=self._nc)  \n",
+    "        else:\n",
+    "            torch.zeros(1)\n",
+    "\n",
+    "        return mp, mr, map50, map\n",
+    "\n",
+    "    def update(self, output, target):\n",
+    "        \"\"\" Calculates and updates metric value\n",
+    "        :param output: model output\n",
+    "        :param target: annotations\n",
+    "        \"\"\"\n",
+    "\n",
+    "        annotation = target[0][\"target\"]\n",
+    "        width = target[0][\"width\"]\n",
+    "        height = target[0][\"height\"]\n",
+    "        shapes = target[0][\"shape\"]\n",
+    "        paths = target[0][\"image_path\"]\n",
+    "        im = target[0][\"img\"]\n",
+    "\n",
+    "        iouv = torch.linspace(0.5, 0.95, 10).to(self._device)  # iou vector for mAP@0.5:0.95\n",
+    "        niou = iouv.numel()\n",
+    "        seen = 0\n",
+    "        stats = []\n",
+    "        # NMS\n",
+    "        annotation = torch.Tensor(annotation)\n",
+    "        annotation[:, 2:] *= torch.Tensor([width, height, width, height]).to(self._device)  # to pixels\n",
+    "        lb = []\n",
+    "        out = output[0]\n",
+    "        out = torch.Tensor(out).to(self._device)\n",
+    "        out = non_max_suppression(out, self._conf_thres, self._iou_thres, labels=lb,\n",
+    "                                  multi_label=True, agnostic=self._single_cls)\n",
+    "        # Metrics\n",
+    "        for si, pred in enumerate(out):\n",
+    "            labels = annotation[annotation[:, 0] == si, 1:]\n",
+    "            nl = len(labels)\n",
+    "            tcls = labels[:, 0].tolist() if nl else []  # target class\n",
+    "            _, shape = Path(paths[si]), shapes[si][0]\n",
+    "            seen += 1\n",
+    "\n",
+    "            if len(pred) == 0:\n",
+    "                if nl:\n",
+    "                    stats.append((torch.zeros(0, niou, dtype=torch.bool), torch.Tensor(), torch.Tensor(), tcls))\n",
+    "                continue\n",
+    "\n",
+    "            # Predictions\n",
+    "            if self._single_cls:\n",
+    "                pred[:, 5] = 0\n",
+    "            predn = pred.clone()\n",
+    "            scale_coords(im[si].shape[1:], predn[:, :4], shape, shapes[si][1])  # native-space pred\n",
+    "\n",
+    "            # Evaluate\n",
+    "            if nl:\n",
+    "                tbox = xywh2xyxy(labels[:, 1:5])  # target boxes\n",
+    "                scale_coords(im[si].shape[1:], tbox, shape, shapes[si][1])  # native-space labels\n",
+    "                labelsn = torch.cat((labels[:, 0:1], tbox), 1)  # native-space labels\n",
+    "                correct = process_batch(predn, labelsn, iouv)\n",
+    "            else:\n",
+    "                correct = torch.zeros(pred.shape[0], niou, dtype=torch.bool)\n",
+    "            stats.append((correct.cpu(), pred[:, 4].cpu(), pred[:, 5].cpu(), tcls))  # (correct, conf, pcls, tcls)\n",
```

Comment:
I use it because the annotation [tensor](https://github.com/sammysun0711/yolov5/blob/openvino_quantization/openvino_quantization.py#L178) can be created on CUDA GPU (even we hard code to CPU for this demo). The postprcossing part can be down on CPU/Nv GPU, use .cpu() function just make sure the cuda torch Tensor convert to CPU numpy array.

I keep this usage in case OV's plugin future plan, can base this code to extend usage of accelerators.